lib/collection/src/grouping/aggregator.rs

```diff
@@ -0,0 +1,332 @@
+use std::collections::{HashMap, HashSet};
+
+use segment::types::{ExtendedPointId, ScoredPoint};
+
+type Hits = HashSet<ScoredPoint>;
+
+#[allow(dead_code)] // temporary
+struct GroupsAggregator {
+    groups: HashMap<String, Hits>,
+    max_group_size: usize,
+    grouped_by: String,
+    max_groups: usize,
+}
+
+#[allow(dead_code)] // temporary
+impl GroupsAggregator {
+    fn new(groups: usize, group_size: usize, grouped_by: String) -> Self {
+        Self {
+            groups: HashMap::with_capacity(groups),
+            max_group_size: group_size,
+            grouped_by,
+            max_groups: groups,
+        }
+    }
+
+    /// Adds a point to the group that corresponds based on the group_by field, assumes that the point has the group_by field
+    fn add_point(&mut self, point: &ScoredPoint) {
+        // if the key contains multiple values, grabs the first one
+        let group_value = point
+            .payload
+            .as_ref()
+            .unwrap()
+            .get_value(&self.grouped_by)
+            .values()
+            .first()
+            .unwrap()
+            .to_string()
+            .trim_matches('"')
+            .to_owned();
+
+        println!("point_id: {:?}, group_value: {:#?}", point.id, group_value);
+
+        if !self.groups.contains_key(&group_value) && self.groups.len() >= self.max_groups {
```

Comment:
`self.groups.len() >= self.max_groups` could be placed right at the beginning to shortcut the whole execution maybe

Reply:
We could still be receiving a point that's going to be bucketed into an existing group, even if we have enough groups